### PR TITLE
lib/model, lib/config: Refactor folder health/error handling (fixes #4445 #4451)

### LINKED ISSUE #4445: Filesystem watcher doesn't handle restarted folders
I have my photos on an external SSD. It's a folder by itself, and usually not connected. When it is connected it gets picked up by a scan and then things work. FS watching doesn't seem compatible with this.

Expected behavior: watching starts when a scan notices the folder is alive again.

Actual behavior: no watching, ever.

### PR DESCRIPTION
**Pull request by imsodin**
While fixing the two linked bugs the result grew uglier and uglier, so I decided to refactor. This splits code regarding folder error/health handling from model.go into folder.go and the config package. The only change in behaviour regards logging and start/stopping fs watch on error state change.

### CODE REVIEW

File: lib/config/size.go [hunk 1]
```diff
@@ -73,3 +75,24 @@ func (s Size) String() string {
...
+			return fmt.Errorf("insufficient space in %v %v: %f %% < %v", fs.Type(), fs.URI(), freePct, s)
+		}
+	} else {
+		if err == nil && float64(usage.Free) < val {
+			return fmt.Errorf("insufficient space in %v %v: %v < %v", fs.Type(), fs.URI(), usage.Free, s)
+		}
+	}
+
+	return nil
+}
```
Comment (AudriusButkevicius) [inline lib/config/size.go]: All of this should live in `func (f *FolderConfiguration) CheckFreeSpace() (err error) {`, and keep using a type, as a type.

File: lib/config/wrapper.go [hunk 2]
```diff
@@ -468,3 +470,15 @@ func (w *Wrapper) MyName() string {
 	cfg, _ := w.Device(myID)
 	return cfg.Name
 }
+
+// CheckHomeFreeSpace returns nil if the home disk has the required amount of
+// free space, or if home disk free space checking is disabled.
+func (w *Wrapper) CheckHomeFreeSpace() error {
+	w.mut.Lock()
+	minFree := w.cfg.Options.MinHomeDiskFree
+	w.mut.Unlock()
```
Comment (AudriusButkevicius) [inline lib/config/wrapper.go]: Could just call `w.Options()`

File: lib/model/rwfolder.go [hunk 3]
```diff
@@ -214,8 +208,8 @@ func (f *sendReceiveFolder) Serve() {
 				continue
 			}
 
-			if err := f.model.CheckFolderHealth(f.folderID); err != nil {
-				l.Infoln("Skipping pull of", f.Description(), "due to folder error:", err)
+			if err := f.CheckHealth(); err != nil {
+				l.Debugln("Skipping pull of", f.Description(), "due to folder error:", err)
```
Comment (AudriusButkevicius) [inline lib/model/rwfolder.go]: We already log this in info or higher, so I'd just remove this

File: lib/model/model.go [hunk 4]
```diff
@@ -2020,13 +2015,13 @@ func (m *Model) internalScanFolderSubdirs(ctx context.Context, folder string, su
...
-			l.Infof("Stopping folder %s mid-scan due to folder error: %s", folderCfg.Description(), iterError)
+			l.Debugln("Stopping scan of folder %s due to: %s", folderCfg.Description(), err)
 			return iterError
 		}
 	}
 
-	if err := m.CheckFolderHealth(folder); err != nil {
-		l.Infof("Stopping folder %s mid-scan due to folder error: %s", folderCfg.Description(), err)
+	if err := runner.CheckHealth(); err != nil {
+		l.Debugln("Stopping scan of folder %s due to: %s", folderCfg.Description(), err)
```
Comment (AudriusButkevicius) [inline lib/model/model.go]: We already log this in info or higher, so I'd just remove this

File: lib/model/model.go [hunk 5]
```diff
@@ -2020,13 +2015,13 @@ func (m *Model) internalScanFolderSubdirs(ctx context.Context, folder string, su
 		})
 
 		if iterError != nil {
-			l.Infof("Stopping folder %s mid-scan due to folder error: %s", folderCfg.Description(), iterError)
+			l.Debugln("Stopping scan of folder %s due to: %s", folderCfg.Description(), err)
```
Comment (AudriusButkevicius) [inline lib/model/model.go]: We already log this in info or higher, so I'd just remove this

File: lib/model/model.go [hunk 6]
```diff
@@ -1914,7 +1909,7 @@ func (m *Model) internalScanFolderSubdirs(ctx context.Context, folder string, su
 		// The error we get here is likely an OS level error, which might not be
 		// as readable as our health check errors. Check if we can get a health
 		// check error first, and use that if it's available.
-		if ferr := m.CheckFolderHealth(folder); ferr != nil {
+		if ferr := runner.CheckHealth(); ferr != nil {
 			err = ferr
 		}
 		runner.setError(err)
```
Comment (AudriusButkevicius) [inline lib/model/model.go]: Isn't this pointless?

File: lib/model/model.go [hunk 7]
```diff
@@ -1926,8 +1921,8 @@ func (m *Model) internalScanFolderSubdirs(ctx context.Context, folder string, su
 
 	for f := range fchan {
 		if len(batch) == maxBatchSizeFiles || batchSizeBytes > maxBatchSizeBytes {
-			if err := m.CheckFolderHealth(folder); err != nil {
-				l.Infof("Stopping folder %s mid-scan due to folder error: %s", folderCfg.Description(), err)
+			if err := runner.CheckHealth(); err != nil {
+				l.Debugln("Stopping scan of folder %s due to: %s", folderCfg.Description(), err)
```
Comment (AudriusButkevicius) [inline lib/model/model.go]: We already log this in info or higher, so I'd just remove this

File: lib/model/model.go [hunk 8]
```diff
@@ -1938,8 +1933,8 @@ func (m *Model) internalScanFolderSubdirs(ctx context.Context, folder string, su
 		batchSizeBytes += f.ProtoSize()
 	}
 
-	if err := m.CheckFolderHealth(folder); err != nil {
-		l.Infof("Stopping folder %s mid-scan due to folder error: %s", folderCfg.Description(), err)
+	if err := runner.CheckHealth(); err != nil {
+		l.Debugln("Stopping scan of folder %s due to: %s", folderCfg.Description(), err)
```
Comment (AudriusButkevicius) [inline lib/model/model.go]: We already log this in info or higher, so I'd just remove this

File: lib/config/size.go (hunk 1 above)
Comment (imsodin) [reply lib/config/size.go]: This code is also needed to get free space of home.

File: lib/model/model.go (hunk 6 above)
Comment (imsodin) [reply lib/model/model.go]: Not sure, the reason given in the comments seems sound enough to me?

File: lib/model/model.go (hunk 7 above)
Comment (imsodin) [reply lib/model/model.go]: Agreed, the only added information is, that it happened during scan, which should also be clear from context.

File: lib/config/size.go (hunk 1 above)
Comment (AudriusButkevicius) [reply lib/config/size.go]: Fine, so have a utility function or something, don't just wedge something on a type.

File: lib/model/model.go (hunk 6 above)
Comment (AudriusButkevicius) [reply lib/model/model.go]: A right, if it's err not ferr, it wouldn't be set otherwise.

File: lib/model/model.go (hunk 7 above)
Comment (AudriusButkevicius) [reply lib/model/model.go]: you can do CheckHealth("during scan") or something to explain where the health check is happening, and then log that.

File: lib/config/size.go (hunk 1 above)
Comment (imsodin) [reply lib/config/size.go]: Wedging was more or less a random decision, I just thought this makes sense here as it is used in the config package (before it was wedged on `Model`, but it has no connection to that at all). Would it make more sense as a utility function in the fs package?

Comment (AudriusButkevicius) [reply lib/config/size.go]: I think so.

Comment (imsodin) [reply lib/config/size.go]: config fs import cycle...
So either we leave it in config or I move the entire Size thing into the fs package.

Comment (AudriusButkevicius) [reply lib/config/size.go]: Make a utility in config.

File: lib/model/model.go (hunk 7 above)
Comment (imsodin) [reply lib/model/model.go]: Actually, we do not necessarily log anything on info or higher unless the error is new or changed. So in a debug scenario, I think it is valid that we log the error anyway and from the position of origin.

Comment (AudriusButkevicius) [reply lib/model/model.go]: Fine whatever.

**Comment by AudriusButkevicius** [comment]
@st-review lgtm

**Comment by st-review** [comment]
@AudriusButkevicius: Noted! Need another LGTM or explicit merge command.

**Comment by calmh** [comment]
@st-review lgtm

lib/model, lib/config: Refactor folder health/error handling (fixes #4445, fixes #4451)

**Comment by st-review** [comment]
:ok_hand: Merged as dc42db444b17145b7de9a0d1c56a42b9d64f3dea. Thanks, @imsodin!

**Pull request closed without merging**